core/drivers/clk/clk.c

```diff
@@ -310,3 +321,108 @@ TEE_Result clk_get_rates_array(struct clk *clk, size_t start_index,
 
 	return clk->ops->get_rates_array(clk, start_index, rates, nb_elts);
 }
+
+/* Return updated message buffer position of NULL on failure */
+static __printf(3, 4) char *add_msg(char *cur, char *end, const char *fmt, ...)
+{
+	va_list ap = { };
+	int max_len = end - cur;
+	int ret = 0;
+
+	va_start(ap, fmt);
+	ret = vsnprintf(cur, max_len, fmt, ap);
+	va_end(ap);
+
+	if (ret < 0 || ret >= max_len)
+		return NULL;
+
+	return cur + ret;
+}
+
+static void __maybe_unused print_clock(struct clk *clk, int indent)
+{
+	static const char * const rate_unit[] = { "Hz", "kHz", "MHz", "GHz" };
+	int rate_unit_index = 0;
+	unsigned long rate = 0;
+	char msg_buf[128] = { };
+	char *msg_end = msg_buf + sizeof(msg_buf);
+	char *msg = msg_buf;
+	int i = 0;
+
+	/*
+	 * Current prints the clock state based on the clock refcount.
+	 * A futur change could print the hardware clock state when
+	 * related clock driver provides a struct clk_ops::is_enabled handler
+	 */
+
+	if (indent) {
+		for (i = 0; i < indent - 1; i++) {
+			msg = add_msg(msg, msg_end, "|   ");
+			if (!msg)
+				goto out;
+		}
+
+		msg = add_msg(msg, msg_end, "+-- ");
+		if (!msg)
+			goto out;
+	}
+
+	rate = clk_get_rate(clk);
+	if (rate == (rate / 1000 * 1000)) {
+		if (rate != (rate / 1000000) * 1000000) {
+			rate_unit_index = 1;
+			rate /= 1000;
+		} else if (rate != (rate / 1000000000) * 1000000000) {
+			rate_unit_index = 2;
+			rate /= 1000000;
+		} else {
+			rate_unit_index = 3;
+			rate /= 1000000000;
+		}
+	}
+
+	msg = add_msg(msg, msg_end, "%s \t(%3s / refcnt %u / %ld %s)",
+		      clk_get_name(clk),
+		      refcount_val(&clk->enabled_count) ? "on " : "off",
+		      (bool)refcount_val(&clk->enabled_count),
+		      rate, rate_unit[rate_unit_index]);
+	if (!msg)
+		goto out;
+
+out:
+	if (!msg)
+		snprintf(msg_end - 4, 4, "...");
+
+	IMSG("%s", msg_buf);
+}
+
+static void print_clock_tree(struct clk *clk_root __maybe_unused,
+			     int indent __maybe_unused)
+{
+#ifdef CFG_DRIVERS_CLK_SUMMARY
+	struct clk *clk = NULL;
+
+	STAILQ_FOREACH(clk, &clock_list, link) {
+		if (clk_get_parent(clk) == clk_root) {
+			print_clock(clk, indent + 1);
+			print_clock_tree(clk, indent + 1);
+			if (indent == -1)
+				IMSG("%s", "");
```

Comment:
For those who wonder why `IMSG("%s", "")` and not a straight use of `IMSG("");`, that is because of some compilers that complained on use of a printf-like function with an empty string argument. I'd rather not have this piece of code to produce a warning on some builds (buildroot, Yocto, ...).